hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java

```diff
@@ -257,40 +266,117 @@ private static boolean applyConstraints(
 			PropertyDescriptor propertyDesc,
 			Set<Class<?>> groups,
 			boolean canApplyNotNull,
+			boolean useOrLogicForComposedConstraint,
+			Dialect dialect) {
+
+		if ( constraintDescriptors.isEmpty() ) {
+			return false;
+		}
+
+		Iterator<ConstraintDescriptor<?>> descriptorIterator = constraintDescriptors.iterator();
+		ConstraintDescriptor<?> descriptor = descriptorIterator.next();
+
+		// head
+		boolean hasNotNull = applyConstraint(
+				descriptor,
+				property,
+				propertyDesc,
+				groups,
+				canApplyNotNull,
+				dialect
+		);
+
+		// tail
+		BiFunction<Boolean, Boolean, Boolean> compositionLogic = useOrLogicForComposedConstraint ?
+				Boolean::logicalAnd :
+				Boolean::logicalOr;
+		while (descriptorIterator.hasNext()) {
+			descriptor = descriptorIterator.next();
+			boolean isNotNull = applyConstraint(
+					descriptor,
+					property,
+					propertyDesc,
+					groups,
+					canApplyNotNull,
+					dialect
+			);
+			hasNotNull = compositionLogic.apply( hasNotNull, isNotNull );
+		}
```

Comment:
All this looks unnecessarily complicated to me. I don't see why we need to use a function reference here, and I don't see why this can't just be written as a `for` loop.

Reply:
Fair. Updated.